### PR TITLE
Refactor Auth server to use wow_login_messages

### LINKED ISSUE #35: Refactor Auth server to use wow_messages.
The world server has been fully reworked to use wow_messages but the auth server still contains custom reading/writing code. Refactor the auth server to use [wow_login_messages](https://github.com/gtker/wow_messages/tree/main/wow_login_messages)

### PR DESCRIPTION
**Pull request by bigglesss**
Fixes #35

Needs some cleanup, and is currently pointed at a local path instead of the wow_login_messages crate, as v0.1 doesn't seem to have async features.

### CODE REVIEW

File: auth_server/src/realms.rs [hunk 1]
```diff
@@ -61,6 +65,39 @@ pub async fn receive_realm_pings(auth_db: std::sync::Arc<AuthDatabase>) -> Resul
     }
 }
 
+async fn get_realm_list(auth_database: std::sync::Arc<AuthDatabase>, account_id: u32) -> Result<Vec<Realm>> {
+    //TODO(wmxd): it will be good idea to cache the database stuff
+    //TODO(wmxd): for now it will be better select realms and number_of_chars in one database trip (eg: left join)
```
Comment (bigglesss) [inline auth_server/src/realms.rs:70]: Going to do this before merging, should be simple enough

File: auth_server/src/realms.rs [hunk 2]
```diff
@@ -61,6 +65,39 @@ pub async fn receive_realm_pings(auth_db: std::sync::Arc<AuthDatabase>) -> Resul
...
+    let db_realms = auth_database.get_all_realms().await?;
+    let mut realms = Vec::with_capacity(db_realms.len());
+    for realm in db_realms {
+        let num_characters = auth_database.get_num_characters_on_realm(account_id, realm.id).await?;
+
+        // TODO: Use flags from DB.
+        let mut flag = Realm_RealmFlag::empty();
```
Comment (bigglesss) [inline auth_server/src/realms.rs:77]: `Realm_RealmFlag` doesn't have a `from` impl for converting from the inner u8 yet. If it did, this whole chunk of code could be much simpler.

File: auth_server/src/realms.rs [hunk 3]
```diff
@@ -61,6 +65,39 @@ pub async fn receive_realm_pings(auth_db: std::sync::Arc<AuthDatabase>) -> Resul
...
+
+        let realm_type: RealmType = RealmType::try_from(realm.realm_type).unwrap_or_default();
+
+        realms.push(Realm {
+            realm_type,
+            locked: 0,
+            flag,
+            name: realm.name,
+            address: realm.ip,
+            population: Population::GreenRecommended,
```
Comment (bigglesss) [inline auth_server/src/realms.rs:91]: The Population in your messages/db seems to be an f32 which is population_count / MAX_POPULATION. I'm not sure how to convert that into a u32 for the message, or if the 3.3.5 client is actually meant to receive an f32?

Comment (bigglesss) [reply auth_server/src/realms.rs:91]:
```
Gtker — Today at 11:39 AM
It is technically an f32, but wowm doesn't support floats for enums, so I just decided to hardcode the float values as u32
It determines which servers are shown as high/med/low pop, although I don't really know what the algorithm is
```

**Notes from looking at other emulators**

Vmangos:

```c
        float popu = float(GetActiveSessionCount());        // updated number of users on the server
        popu /= pLimit;
        popu *= 2;

        SqlStatement stmt = LoginDatabase.CreateStatement(id, "UPDATE `realmlist` SET `population` = ? WHERE `id` = ?");
```

Always sends a float based on calculation, never hardcoded `Recommended` values.

File: auth_server/src/realms.rs (hunk 2 above)
Comment (Victov) [reply auth_server/src/realms.rs:77]: I usually enter a GH issue with a feature request on the wowm repo and then put a link in the comment like "TODO refactor this when this GH issue is fixed"

File: auth_server/src/realms.rs (hunk 1 above)
Comment (bigglesss) [reply auth_server/src/realms.rs]: It doesn't seem like SQLX is very good at handling `joins` in a nice re-usable way. If you have a look at what I've done in d3be028f7b45e2259c6f0d0483369301c6544fa1, you'll see that I had to duplicate `DBRealm` and add the `num_characters` column as an `Option` for the left join. Super annoying.

I thought about just having the `Option` in the original DBRealm and adding a literal like `NULL as num_characters` or `0 as num_characters` to the normal `get_all_realms` query, but it doesn't seem like there's a good way of defining a nullable literal in an sql query. Maybe the answer is just to use `query!` for joins and keep them as anonymous records, but then they can't be returned from functions.

File: auth_server/src/realms.rs (hunk 3 above)
Comment (bigglesss) [reply auth_server/src/realms.rs]: I've changed this logic to convert the f32 into the u32 equivalent representation of the f32, so the client should be able to parse it correctly as an f32 (similar to how the normal enum values are handled).

File: auth_server/src/realms.rs (hunk 1 above)
Comment (Victov) [reply auth_server/src/realms.rs]: Having both `DBRealm` and `DBRealmWithNumCharacters` is totally fine by me.

**Pull request merged**